File: all/226_invert_binary_tree/sol_postorder.py
from typing import Optional


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        def invertTreeByPostorderTraversal(root: Optional[TreeNode]) -> None:
            if not root:
                return

            # Invert the left subtree and the right subtree first.
            invertTreeByPostorderTraversal(root.left)
            invertTreeByPostorderTraversal(root.right)

            # Swap the left subtree and the right subtree.
            (root.left), (root.right) = (root.right), (root.left)

        invertTreeByPostorderTraversal(root)

        return root

File: all/226_invert_binary_tree/sol_postorder.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
void invertTreeByPostorderTraversal(struct TreeNode * const root) {
    struct TreeNode *original_left_subtree_ptr = NULL;

    if (NULL == root) {
        return;
    }

    invertTreeByPostorderTraversal(root->left);
    invertTreeByPostorderTraversal(root->right);

    /* Swap the left subtree and right subtree. */
    original_left_subtree_ptr = root->left;
    root->left = root->right;
    root->right = original_left_subtree_ptr;

    return;
}

struct TreeNode* invertTree(struct TreeNode* root) {
    /*
     * User recursive postorder traversal to swap the left subtree and
     * right subtree of a given root node.
     */
    invertTreeByPostorderTraversal(root);

    return root;
}
